listenbrainz/db/year_in_music.py

```diff
@@ -41,111 +39,129 @@
-             SELECT "user".id
-                  , {year}
-                  , jsonb_build_object({key}, data::jsonb)
-               FROM (VALUES %s) AS t(user_id, data)
-               JOIN "user"
-                 ON "user".id = user_id::int
-        ON CONFLICT (user_id, year)
-      DO UPDATE SET data = COALESCE(statistics.year_in_music.data, '{{}}'::jsonb) || EXCLUDED.data
-    """).format(key=Literal(key), year=Literal(year))
-    try:
-        with connection.cursor() as cursor:
-            values = [(user["user_id"], orjson.dumps(user["data"]).decode("utf-8")) for user in data]
-            execute_values(cursor, query, values)
-        connection.commit()
-    except psycopg2.errors.OperationalError:
-        connection.rollback()
-        current_app.logger.error("Error while inserting top stats:", exc_info=True)
+def insert_heavy(key, year, data):
+    insert(key, year, [(user["user_id"], orjson.dumps(user["data"]).decode("utf-8")) for user in data], True)
 
 
-def handle_insert_top_stats(entity, year, data):
-    connection = db.engine.raw_connection()
-    query = SQL("""
-        INSERT INTO statistics.year_in_music (user_id, year, data)
-             SELECT "user".id
-                  , {year}
-                  , jsonb_build_object({key}, data::jsonb, {count_key}, count)
-               FROM (VALUES %s) AS t(user_id, count, data)
-               JOIN "user"
-                 ON "user".id = user_id::int
-        ON CONFLICT (user_id, year)
-      DO UPDATE SET data = COALESCE(statistics.year_in_music.data, '{{}}'::jsonb) || EXCLUDED.data
-    """).format(key=Literal(f"top_{entity}"), count_key=Literal(f"total_{entity}_count"), year=Literal(year))
-    try:
-        with connection.cursor() as cursor:
-            values = [(user["user_id"], user["count"], orjson.dumps(user["data"]).decode("utf-8")) for user in data]
-            execute_values(cursor, query, values)
-        connection.commit()
-    except psycopg2.errors.OperationalError:
-        connection.rollback()
-        current_app.logger.error("Error while inserting top stats:", exc_info=True)
+def insert_similar_users(year, data):
+    user_ids = get_all_usernames()
+
+    similar_user_data = [
+        {
+            "user_id": int(user_id),
+            "data": orjson.dumps(
+                [
+                    {
+                        "musicbrainz_id": user_ids[int(other_user_id)],
+                        "score": score
+                    }
+                    for other_user_id, score in similar_users.items()
+                    if int(other_user_id) in user_ids
+                ]
+            ).decode("utf-8")
+        }
+        for user_id, similar_users in data.items()
+    ]
+
+    insert_heavy("similar_users", year, similar_user_data)
+
+
+def insert_top_stats(entity, year, data):
+    insert_heavy(f"top_{entity}", year, data)
+    insert(f"total_{entity}_count", year, [(user["user_id"], user["count"]) for user in data], False)
```

Comment:
I think we can improve this by using on INSERT statement, but lets skip it for this second.

Reply:
Sorry I am not sure what you mean?

Reply:
ah okay, I see. Yes that is possible, can do later.